Bugfixes and features

okahilak opened the pull request:
List of added features:

- When creating coil target from a landmark, move it away from scalp by 1 cm as a default. This seems to be what the experimenter usually does anyway, as they don't want to risk creating the target too close to the subject's head.
- Make texts in buttons more descriptive when creating a peeled surface
- Add Z-offset column to marker list, indicating how many millimeters the coil target has been moved closer to/away from the head
- Improve phrasing of error messages in image/tracker tabs
- Try to retain coil z-rotation when moving markers on scalp. Not perfect; currently only retains the rotation that has been applied to the marker using keyboard hotkeys, not the initial rotation that a coil target has when it's created from an actual coil position.
- Store sleep parameters (under "Navigation" tab in preferences) in config, preventing the need to reset them when InVesalius is restarted
- Store distance and angle thresholds in config, preventing the need to reset them when InVesalius is restarted.
- Select a marker in the volume viewer by left mouse click
- Double-click on marker in the volume viewer to focus on it, making the camera rotate around it - this helps to visually observe, e.g., the distance of marker to the scalp.
- Double-click on a surface in the volume viewer to focus on it - this allows resetting the camera movement so that it rotates around the scalp or the brain, which is what we usually want.
- Double-click on a marker in the marker list to focus on it in the volume viewer, making it easier to find where the marker is located.
- When a selected marker is deleted in the marker list, automatically select the first remaining marker on the list; for instance, if marker no 3 is selected and deleted, automatically select the new marker no 3 (which was previously marker no 4).
- Automatically select and focus on a newly-created marker. For instance, when a coil target is created from a landmark, it is now automatically selected on the marker list, as well as highlighted in the volume viewer.
- Automatically reconnect to the tracker if same coordinates are received repeatedly; this is useful in case of OptiTrack, which occasionally goes into a failure mode where it keeps sending the same coordinates. Reconnecting resolves the issue. However, this is a workaround and should be eventually fixed properly, e.g., by updating OptiTrack drivers.
- Enable robot's "move away from the head" button even when a target is not set. This is convenient for initializing the physical set-up of the system while a target has not yet been found or decided.
- Display an error if head or probe tracker is not visible during the registration of the subject.
- Display an error if coil or probe tracker is not visible during the registration of the coil.

Changes to existing features:

- Use smoothed scalp surface when moving markers on scalp. Using the raw surface caused the markers to have weird orientations at times, due to the uneven surface of the scalp mesh. Using smoothed surface alleviates that issue. Compute the smoothed surface on the fly when needed.
- Use more convenient step size when moving markers; e.g., increment/decrement the rotation angle by 5 deg instead of 1 deg.
- When in target mode, always move the target marker, not the marker selected in the marker list, when using keyboard hotkeys (A, D, etc.). The previous behavior caused some confusion during an experiment, when the experimenter wanted to move the target, but another marker was accidentally selected on the list.
- Spin the camera in volume viewer instead of zooming when right mouse button is pressed and mouse is moved over the viewer; this decreases the redundancy of different mouse button functions, as the scroll-wheel can already be used for zooming, and more precise zooming can be still done by selecting the 'zoom' toolbar item. Spinning can be used, e.g., to get a view where camera is targeted down towards the brain so that the anterior is at the top and posterior is at the bottom of the view.
- Focus on marker when it's moved using keyboard hotkeys - this helps in keeping the marker in view while it's being moved.
- Increase horizontal size of several dialog boxes, namely, "Load markers" dialog and tracker configuration dialog boxes, which were too narrow previously to fully display the selected path and the filename.
- Increase height of the marker list; previously, it showed four markers at a time. After the change, if shows 10 markers, which is more convenient and user-friendly.
- Do not enter target mode automatically when target is being moved; it is typically not what we want, as it doesn't allow the target to be moved repeatedly outside the target mode.
- Do not auto-focus on slice viewers when mouse is moved over them; auto-focusing is convenient when operating with slice viewers, but it noticeably decreases the usability of the marker list and the volume viewer, as it often causes the slice viewers to inadvertently grab the focus. In particular, it makes the interplay between marker list and the volume viewer difficult for the user, as auto-focusing makes it hard to, e.g., select a marker in the marker list, move the cursor to the volume viewer to rotate and move it around, while keeping the marker selected and responsive to, e.g., hotkeys to move the marker.

Bugfixes:

- "Back" button on "refine" page moved to incorrect page during registration
- Highlighting a marker did not work properly when it had exact duplicates due to the rendering order only allowing the last-rendered marker to be visible, preventing the others from being highlighted.
- Marker was unhighlighted when it was clicked twice on the marker list
- 'Distance to target' text was not always visible in the target mode
- Ensure only one marker is highlighted at a time; previously, the attempt to highlight several markers, which is not currently supported, caused some visual glitches.
- Selecting multiple markers, needed for deleting several markers at once, slowed down InVesalius a lot, due to issues in highlighting the markers.

okahilak commented:
Rebased these changes on top of #761. The changes in this PR can be reviewed separately and merged independently of #761; once the latter is merged, I think this PR will automatically update itself to show only the new commits.

rmatsuda commented:
Thanks for the pull request @okahilak! I tried to use the keyboard to move the target, the rotation is working fine but the translation is returning the following error:

```
D:\Repository\invesalius3-main\invesalius\data\markers\marker_transformator.py:199: RuntimeWarning: invalid value encountered in scalar divide
  rotation_angle = np.arccos(np.dot(ref_vector, normal_vector) / (np.linalg.norm(ref_vector) * np.linalg.norm(normal_vector)))
D:\Repository\invesalius3-main\invesalius\data\markers\marker_transformator.py:202: RuntimeWarning: invalid value encountered in divide
  rotation_axis_normalized = rotation_axis / np.linalg.norm(rotation_axis)
Traceback (most recent call last):
  File "D:\Repository\invesalius3-main\invesalius\gui\frame.py", line 185, in OnGlobalKey
    Publisher.sendMessage('Move marker by keyboard', keycode=keycode)
  File "D:\Repository\invesalius3-main\invesalius\pubsub\pub.py", line 74, in sendMessage
    Publisher.sendMessage(topicName, **msgdata)
  File "C:\Program Files\anaconda3\envs\invesalius-cp311\Lib\site-packages\pubsub\core\publisher.py", line 216, in sendMessage
    topicObj.publish(**msgData)
  File "C:\Program Files\anaconda3\envs\invesalius-cp311\Lib\site-packages\pubsub\core\topicobj.py", line 452, in publish
    self.__sendMessage(msgData, topicObj, msgDataSubset)
  File "C:\Program Files\anaconda3\envs\invesalius-cp311\Lib\site-packages\pubsub\core\topicobj.py", line 482, in __sendMessage
    listener(data, self, allData)
  File "C:\Program Files\anaconda3\envs\invesalius-cp311\Lib\site-packages\pubsub\core\listener.py", line 237, in __call__
    cb(**kwargs)
  File "D:\Repository\invesalius3-main\invesalius\data\markers\marker_transformator.py", line 308, in MoveMarkerByKeyboard
    self.MoveMarkerOnScalp(
  File "D:\Repository\invesalius3-main\invesalius\data\markers\marker_transformator.py", line 132, in MoveMarkerOnScalp
    self.ProjectToScalp(
  File "D:\Repository\invesalius3-main\invesalius\data\markers\marker_transformator.py", line 178, in ProjectToScalp
    closest_point, closest_normal = self.surface_geometry.GetClosestPointOnSurface('scalp', marker_position)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Repository\invesalius3-main\invesalius\data\markers\surface_geometry.py", line 178, in GetClosestPointOnSurface
    closest_normal = normal_data.GetTuple(closest_point_id)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: expects 0 <= tupleIdx && tupleIdx < GetNumberOfTuples()
```

okahilak commented:
Thanks for testing and reporting this, @rmatsuda!

I wasn't able to reproduce the error using the example project or another project that I've been testing on. Could you ensure that you can reproduce it when you start a fresh InVesalius instance, open a saved project and a set of previously-stored markers, and start moving a marker in that project? Could you then send me that project and the markers that you are using? A .zip file into my e-mail works fine for that purpose.

Thanks a lot!

rmatsuda commented:
Thanks, @okahilak! I started a fresh instance and created a new InVesalius project. It's working fine! But after I closed and restored the session, it's returning the following error and all surfaces disappear from 3D:
```
D:\Repository\invesalius3-main\invesalius\data\markers\marker_transformator.py:199: RuntimeWarning: invalid value encountered in scalar divide
  rotation_angle = np.arccos(np.dot(ref_vector, normal_vector) / (np.linalg.norm(ref_vector) * np.linalg.norm(normal_vector)))
D:\Repository\invesalius3-main\invesalius\data\markers\marker_transformator.py:202: RuntimeWarning: invalid value encountered in divide
  rotation_axis_normalized = rotation_axis / np.linalg.norm(rotation_axis)
Target updated to coordinates [nan, nan, nan, nan, nan, 0.0]

vhosouza commented:
The error reported by @rmatsuda also happens with me. To reproduce:
1. Open an existing project
2. Create a coil target
3. Restart invesalius by recovering the state
4. Create a new marker using the right click on the 3D surface
5. Set the created target as coil target (it takes a really long time, minutes, to smooth the surface)
6. The 3d scene will be empty the error Renan shared will be raised

An additional bug is that:
1. Dectivating the targeting button/interface, the 3d scene is recovered
2. Activating the targeting button/interface, the 3d scene disappears and the error is raised in the terminal

Below a couple of screenshots of the additional bug:

![invesalius_bug_target_2](https://github.com/invesalius/invesalius3/assets/1081406/09a136e5-da0a-4b95-b45b-6e0d6fde76ae)
![invesalius_bug_target_1](https://github.com/invesalius/invesalius3/assets/1081406/80b97cbf-ef1f-47f1-9821-ba80a0333baf)

vhosouza commented:
Although the automatic z offset seem to be intuitive, it is very application specific and in this case can cause unwanted problems. The fact that in Aalto an offset is needed is because the 5coil is being registered on the bottom surface, which does not correspond to the center of the Figure-8 coil 3D object in invesalius which causes a visual displacement. We should fix the coil registration process and alignment with the 3d object and not apply an offset.

The default behavior of creating markers should be create a marker where the click was made without any offset. Also, the 1 cm offset is quite big and has no correspondence to general applications of TMS / neuronavigation. Here is a screenshot of how big difference from the selected marker:

![invesalius_bug_zoffset_1](https://github.com/invesalius/invesalius3/assets/1081406/c765e2d3-4f94-4834-be7a-c195e8f8e67e)

vhosouza commented:
I'm not sure if this was added in this PR or the latest one, but the Figure-8 coil **with handle** should not be used as the target "ghost coil". This makes the 3d scene unnecessarily cluttered and does not bring additional visual guidance. @okahilak if you changed this, please revert, if not, I will ask Asta or Henrik to do it.

vhosouza commented:
There seems to be a very tiny coordinate system (x,y,z) axes in the middle of the orange donut representing the coil center. I suggest removing it.

![invesalius_bug_disc_1](https://github.com/invesalius/invesalius3/assets/1081406/55991357-dcec-4f82-8fa4-4dac3306a76e)

vhosouza commented:
Increasing the size of the markers list is **EXCELLENT**! However, in a smaller laptop screen the robot buttons stay hidden and cannot be accessed.

I suggest adjusting the positioning to the button elements in a way to make everything fit considering a laptop screen.

![invesalius_bug_ui_1](https://github.com/invesalius/invesalius3/assets/1081406/29fd9249-fd22-4a61-9625-0df7af549cac)

okahilak commented:
> The error reported by @rmatsuda also happens with me. To reproduce:
> 
> 1. Open an existing project (also happens when creating a new one)
> 2. Create a coil target
> 3. Start and stop navigation (not sure if this is really needed)
> 4. Create a new marker using the right click on the 3D surface
> 5. Set the created target as coil target (it takes a really long time, minutes, to smooth the surface)
> 6. The 3d scene will be empty the error Renan shared will be raised

Thanks for listing the detailed steps to reproduce the issue! Unfortunately, I'm still unable to reproduce it with the two projects that I have access to. Like you mentioned, it's probably related to smoothing the scalp surface; one possibility is that the smoothing erodes the whole surface, leaving an empty mesh, which then causes an error. However, we didn't encounter that in Aalto during several experiments, all with different surface meshes, so it doesn't seem to happen with all procedures for creating the surfaces.

To investigate further, I would probably need an access to a project with which that issue happens.

okahilak commented:
If sharing the project is unattainable, I would propose testing it after modifying L53-54 in `surface_geometry.py`. Currently they look like this:

```
        smoother.SetNumberOfIterations(400)
        smoother.SetRelaxationFactor(0.9)
```

Just to double-check that it's indeed caused by the smoothing, I'd propose changing the # of iterations to 0, which should make the smoothed surface equivalent to the original. If that resolves the issue, I would keep it at 0 for the time being until we can get a better understanding of how the smoothing should be done so that it always works. (I can make the commit if you confirm that it indeed resolves the problem.)

One more point that comes to mind: for the surfaces that I've tested it with, the smoothing happens in 10-20 seconds with my not-so-powerful laptop. That's another indication that my test projects are using quite different kind of surfaces, compared to yours, and that maybe one-size-fits-all solution cannot be easily found here.

okahilak commented:
@vhosouza, @rmatsuda, thanks for the comments. I did the following changes:

- Reverted the change that moves the coil target 1 cm away from the scalp surface when created from a landmark.
- Deleted x-, y-, and z-axis indicators (the small red, green, and blue lines) when navigating.
- Ensured that all buttons in the navigation panel are visible when using 1080p resolution, which is the most common resolution on laptops. The marker list now becomes taller to occupy as much space as there is available. However, this should be tested with a larger resolution; I currently only have access to a computer with 1080p resolution.
- Target coil mesh doesn't show coil handle now when figure-of-8 coil is used.
- Marker list did not correctly show the target; this is fixed now.

Most importantly, I effectively disabled the scalp smoothing for now. However, the infrastructure remains there, so it can be taken into use in the future. @rmatsuda, could you test if this fixes the issues that you had? If so, we should probably keep it disabled for the time being and investigate later how to make it work consistently.

Here's how the navigation panel currently looks with 1080p resolution. As can be seen in the screenshot, there is no extra space to make the marker list taller than it was previously. However, with larger resolutions, the marker list should automatically occupy the available extra space:

![image](https://github.com/invesalius/invesalius3/assets/49641103/20a953ee-bfca-44d0-b58b-ff560cf3f100)

okahilak commented:
@rmatsuda, @vhosouza, is there any update on this PR and what's still needed to get it merged?

vhosouza commented:
@okahilak sorry, had no time so far to check it again. I'll try to do it soon. If @rmatsuda also has something else to comment, feel free.

vhosouza commented:
@tfmoraes and @okahilak, this is a large set of changes that may affect some basic functionalities of invesalius that others use outside neuronavigation and that @rmatsuda and I wouldn't be aware of. We checked that for the neuronavigation it seems to work fine and is basically ready for merging. But I'd like either of you to do a quick review and test to see if some basic things you do are affected or if you can identify any bug. This is a bit urgent because other developments are depending on this, so please consider it with a high priority.

okahilak commented:
Thanks @vhosouza - just a clarification that you probably intended to ping @paulojamorim.

vhosouza commented:
As discussed with @rmatsuda and @tfmoraes, PR is merged and if bugs appear they should be fixed in smaller PRs.
